Update VecGeom support for 2.0 solid and surface

mrguilima opened the pull request:
Creating this (draft) PR as suggested yesterday at the standup meeting.

I tried today to merge in the develop branch, and then it does not compile.  However, this was passing all tests (including orange, e.g. VecGeom_DIR undefined).

For surface model, tested against VecGeom commit c3ee06912.
For solids model, tested agains VecGeom commit 7b8b07ba4 (then the head of v1-patches).

All tests also build ok for solids model, against the same commit above used for the surface model (c3ee06912), however the solids model fail several tests.  Overall, the solids model currently looks better (e.g. has better agreement agains surface model and Orange) using v1-patches than using VecGeom's master branch.

mrguilima committed:
Partial fix to FieldPropagator tests

mrguilima committed:
Cleanup FieldPropagator test

mrguilima committed:
Fixes to tests after VecGeom surface model merged into master

mrguilima committed:
Test cleanup

mrguilima committed:
Fixes to Geometry and FieldPropagator tests

mrguilima committed:
Merge branch 'develop' into vecgeom-2.0

mrguilima commented:
Updating test results here, after recent updates in VecGeom (v2.x = commit 14cd3a877 in branch 'stable_grazing') and in Celeritas (commit 6ab096711 in branch vecgeom-2.0):

### Using Orange:         100% tests passed,  0 tests failed out of 300  (143.78 sec)

### VecGeom v1.2.10:    99% tests passed,  3 tests failed out of 327  (123.00 sec) <-- solid model only
  The following tests FAILED:
	204 - celeritas/em/UrbanMsc (Failed)                    unit
	205 - celeritas/em/WentzelVIMsc (Failed)                unit
	215 - celeritas/ext/RootJsonDumper (Failed)             unit

### VecGeom v1.x (head of v1-patches):     100% tests passed,  0 tests failed out of 327  (145.72 sec) <-- solid model only

### Using vecgeom-2.x-surf:    99% tests passed, 2 tets failed out of 324  (155.67 sec)     <-- surface model
The following tests FAILED:
	251 - celeritas/global/StepperGeant:TestEm3MscNofluct.* (Failed) gpu unit
	252 - celeritas/global/StepperGeant:TestEm3MscNoIntegral.* (Failed) gpu unit

### VecGeom v2.x-solid:
  The following tests FAILED:
	 89 - geocel/vg/Vecgeom:SimpleCmsVgdmlTest.* (Failed)   gpu unit
	 97 - geocel/vg/Vecgeom:ReplicaTest.* (Failed)          gpu unit
	204 - celeritas/em/UrbanMsc (Failed)                    unit
	205 - celeritas/em/WentzelVIMsc (Failed)                unit
	215 - celeritas/ext/RootJsonDumper (Failed)             unit
	224 - celeritas/field/FieldPropagator:-Layers*:SimpleCms*:Cmse* (Failed) unit
	226 - celeritas/field/FieldPropagator:SimpleCms* (Failed) unit
	227 - celeritas/field/FieldPropagator:Cmse* (Failed)    unit
	230 - celeritas/geo/Geometry:SimpleCms* (Failed)        gpu unit
	231 - celeritas/geo/Geometry:TestEm3* (Failed)          gpu unit
	232 - celeritas/geo/Geometry:ThreeSpheres* (Failed)     gpu unit
	251 - celeritas/global/StepperGeant:TestEm3Compton.* (Failed) gpu unit
	252 - celeritas/global/StepperGeant:TestEm3NoMsc.* (Failed) gpu unit
	253 - celeritas/global/StepperGeant:TestEm3Msc.* (Failed) gpu unit
	254 - celeritas/global/StepperGeant:TestEm3MscNofluct.* (Failed) gpu unit
	304 - celeritas/track/TrackSort (Failed)                gpu unit
	305 - celeritas/track/TrackInit (Failed)                gpu unit
	309 - celeritas/user/SlotDiagnostic (Failed)            gpu unit
	312 - celeritas/user/StepCollector:TestMultiEm3* (Failed) gpu unit
	322 - app/celer-g4:gpu (Timeout)                        app gpu nomemcheck

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vecgeom-2.0

sethrj committed:
Use vecgeom 2.0 rc.4 and 1.2.11

sethrj committed:
REVERTME: disable non-vecgeom CI

sethrj committed:
fixup! REVERTME: disable non-vecgeom CI

sethrj committed:
Use rc.4 to work around vecgeom cmake bug

mrguilima committed:
Fixes to unit tests for VecGeom solids model

mrguilima committed:
Remove redundant code

mrguilima committed:
Disable check on bumps during track()

mrguilima committed:
Relax distance and safety tolerance to 1.e-8

mrguilima committed:
Fix unit tests for Polyhedra and SimpleCms in solids model

mrguilima committed:
Set celer-g4 app testing timeout to 10sec

mrguilima committed:
Set a 12sec timeout for some unit tests

mrguilima committed:
Add SolidsNavigator to unify solids-model interface for v2.x and v1.x

mrguilima committed:
Avoid using relocation at this point, since it is already done on find_next_step()

mrguilima committed:
GL: additional tests to FourLevelsVgdml

mrguilima committed:
Fixes to Vecgeom tests

mrguilima committed:
Merge branch 'develop' into vecgeom-2.0

mrguilima committed:
Some fixes to Vecgeom tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mrguilima committed:
Take back differences in test expectations

mrguilima committed:
Increase app testing TIMEOUT to 20sec

mrguilima committed:
All VecGeom tests fixed on CPU

mrguilima committed:
All celeritas/geo_Geometry tests fixed on GPU

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mrguilima committed:
Fixes to FieldPropagator tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mrguilima committed:
Merge branch 'develop' into vecgeom-2.0

mrguilima committed:
NavIndex_t fix (by Seth)

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

mrguilima committed:
Adjust StepperGeant expected ranges

mrguilima committed:
Loosen up testing timeouts

mrguilima committed:
Merge branch 'develop' into vecgeom-2.0

sethrj committed:
Enable tests, use latest vecgeom+g4vg, use a single vecgeom install

sethrj committed:
Suppress invalid builds

sethrj committed:
Also suspend doc build

sethrj committed:
Choose error message instead of segfault

sethrj reviewed (commented):
@mrguilima Thanks for getting this done! I've added vecgeom 2.0.0-rc.7 to this CI (and re-enabled failing tests), but I am seeing failures.

The failing `accel` tests are because vecgeom surface crashes when we have more than one BeamOn (I've added a commit to this effect). In those accel tests, please have an `if (CELERITAS_VECGEOM_SURFACE) { CELER_LOG(warning) << "VecGeom surface does not support multiple BeamOn";  return;}` before the second BeamOn.

Could you please also look at the other [test failures](https://github.com/celeritas-project/celeritas/pull/1720/checks?check_run_id=52513963237)? Thanks!

sethrj committed:
Restore CI

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vecgeom-2.0

sethrj committed:
Remove builds again

sethrj commented:
@mrguilima Where will this be by the end of today? If necessary we can still leave the 2.0/surface tests disabled but we can't keep wasting time resolving conflicts every week. Please tell me your plans and schedule for the rest of this.

mrguilima committed:
Merge branch 'develop' into vecgeom-2.0

mrguilima committed:
Avoid reloading geometry in tests for VecGeom surface model

mrguilima committed:
Merge branch 'develop' into vecgeom-2.0

mrguilima committed:
Fix OpticalCollector device test

sethrj commented:
@mrguilima Thanks for continuing on this. You're right that there are a lot of combinations. What about we just try to simply fix the [celeritas/field/FieldPropagator:Cmse* test for vecgeom 2 solid](https://github.com/celeritas-project/celeritas/actions/runs/18888421629/job/53909745245?pr=1720), which seems to not be exiting the boundary correctly:
```
warning: Moved internally from boundary but safety didn't increase: volume 18 from {10.31609436324, -6.5649478629707, 796.92284229006} to {10.316160010118, -6.5649741252439, 796.92291300626} (distance: 9.999999997693e-05)
warning: Moved internally from boundary but safety didn't increase: volume 19 from {99.994867290585, -1.0131710311187, 6585.456544248} to {99.74728429231, 4.2955842461032, 6590.7760977579} (distance: 7.519430129917)
warning: Moved internally from boundary but safety didn't increase: volume 19 from {99.74728429231, 4.2955842461032, 6590.7760977579} to {98.706755024331, 9.5072518029854, 6596.0956512728} (distance: 7.5194301299217)
warning: Moved internally from boundary but safety didn't increase: volume 19 from {98.706755024331, 9.5072518029854, 6596.0956512728} to {96.896790709203, 14.504071692219, 6601.4152047927} (distance: 7.5194301299247)
```

This seems to be the same behavior we're seeing [on the develop branch for the slightly older 2.0](https://github.com/celeritas-project/celeritas/actions/runs/18888421629/job/53909745245?pr=1720 ). So maybe we should just leave these disabled and start a new development branch.

sethrj commented on app/celer-g4/CMakeLists.txt:
```diff
@@ -121,6 +121,7 @@ function(celer_app_test ext)
   set_tests_properties("app/celer-g4:${ext}" PROPERTIES
     ENVIRONMENT "${_env};${_extra_env}"
     REQUIRED_FILES "${_required_files}"
+    TIMEOUT 200
```
Please revert these. If you need to reduce the timeout during development, just run `ctest --timeout 200`.

sethrj commented on src/geocel/vg/VecgeomTrackView.hh:
```diff
@@ -563,10 +559,10 @@ CELER_FUNCTION void VecgeomTrackView::cross_boundary()
+        //// Some navigators require an lvalue temp_pos
+        // auto temp_pos = detail::to_vector(this->pos_);
+        // Navigator::RelocateToNextVolume(
+        //     temp_pos, detail::to_vector(this->dir_), vgnext_);
```
If this is no longer needed, please delete the `#else` condition. Also try to make a habit of not using `//` for disabling code. `//` is for textual comments, `#if 0` uses the preprocessor to disable code.

sethrj commented on src/geocel/vg/detail/SolidsNavigator.hh:
```diff
@@ -0,0 +1,246 @@
+{
+namespace detail
+{
+using namespace vecgeom;
```
`using namespace` is prohibited in header files, even inside another namespace. This puts every symbol from vecgeom in `celeritas::detail`.

sethrj commented on src/geocel/vg/detail/SolidsNavigator.hh:
```diff
@@ -0,0 +1,246 @@
+    using Vector3D = vecgeom::Vector3D<vecgeom::Precision>;
+    using VPlacedVolumePtr_t = vecgeom::VPlacedVolume const*;
+
+    // static constexpr Precision kBoundaryPush = 10 * vecgeom::kTolerance;
```
is this no longer needed?

sethrj commented on test/accel/TrackingManagerIntegration.test.cc:
```diff
@@ -411,8 +417,12 @@ TEST_F(LarSphereOptical, run)
+    {
+        CELER_LOG(status) << "Beam on (second run)";
+        rm.BeamOn(2);
+    }
```
I literally gave you the code to use:
```suggestion
    if (using_surface_vg)
    {
        GTEST_SKIP() << "VecGeom surface does not support multiple runs";
    }

    CELER_LOG(status) << "Beam on (second run)";
    rm.BeamOn(2);
```

This changes less code, makes the exceptional case the exception rather than the main path flow, and prints an informative message about why we're skipping the rest of the test. Please change in the other test as well.

sethrj commented on test/celeritas/field/FieldPropagator.test.cc:
```diff
@@ -850,7 +840,7 @@ TEST_F(TwoBoxesTest, electron_corner_hit)
         }
         geo.cross_boundary();
-        EXPECT_EQ("world", this->volume_name(geo));
+        EXPECT_EQ(this->volume_name(geo), "world");
```
Revert please. Originally EXPECT_EQ was defined by googletest as `expected, actual` and we try to keep that convention. Some of the macros (e.g. EXPECT_SOFT_EQ) still identify their arguments as such.

sethrj commented on test/celeritas/field/FieldPropagator.test.cc:
```diff
@@ -1297,14 +1298,18 @@ TEST_F(SimpleCmsTest, TEST_IF_CELERITAS_DOUBLE(electron_stuck))
         }
-        EXPECT_SOFT_EQ(30, calc_radius());
+        // EXPECT_SOFT_EQ(30, calc_radius());
+        EXPECT_SOFT_NEAR(calc_radius(), 29.9999996, 4e-7);
```
That's a substantial accuracy loss for double-precision tracking! Is this expected??

sethrj commented on test/celeritas/field/FieldPropagator.test.cc:
```diff
@@ -1254,30 +1248,38 @@ TEST_F(SimpleCmsTest, TEST_IF_CELERITAS_DOUBLE(electron_stuck))
         EXPECT_EQ(result.boundary, geo.is_on_boundary());
-        if (using_vecgeom_surface)
+
+        EXPECT_EQ("si_tracker", this->volume_name(geo));
+        EXPECT_TRUE(geo.is_on_boundary());
+        EXPECT_FALSE(result.looping);
```
I think most of the following conditionals are redundant now, please check

sethrj commented on test/celeritas/field/FieldPropagator.test.cc:
```diff
@@ -1297,14 +1298,18 @@ TEST_F(SimpleCmsTest, TEST_IF_CELERITAS_DOUBLE(electron_stuck))
+        EXPECT_EQ(this->volume_name(geo),
+                  using_surface_vg ? "vacuum_tube" : "si_tracker");
+        std::cerr << " using surface=" << using_surface_vg
+                  << " solid:" << using_solids_vg << std::endl;
```
Remove debug output

sethrj commented on test/celeritas/field/FieldPropagator.test.cc:
```diff
@@ -1297,14 +1298,18 @@ TEST_F(SimpleCmsTest, TEST_IF_CELERITAS_DOUBLE(electron_stuck))
         EXPECT_SOFT_NEAR(
             double{30}, static_cast<double>(integrate.count()), 0.2);
-        ASSERT_TRUE(geo.is_on_boundary());
+        EXPECT_EQ(geo.is_on_boundary(), !using_surface_vg);
```
You can't call cross_boundary on line 1308 if this isn't true (which is why the test is failing). I'm not sure why this is behaving differently or what the pre-propagation state is supposed to be...

sethrj commented on test/celeritas/field/FieldPropagator.test.cc:
```diff
@@ -1363,12 +1371,43 @@ TEST_F(SimpleCmsTest, TEST_IF_CELERITAS_DOUBLE(vecgeom_failure))
+            // Interestingly, VecGeom surf and solid models see that surface
+            // slightly differently.  Only surface model thinks the surface
+            // was actually crossed, therefore the next step will find distinct
+            // results
```
Great comments to explain the test! 😄

sethrj commented on test/celeritas/global/AlongStep.test.cc:
```diff
@@ -709,14 +711,14 @@ TEST_F(SimpleCmsRZFieldAlongStepTest, msc_rzfield)
-        EXPECT_SOFT_EQ(0.5515596670659112, result.displacement);
-        EXPECT_SOFT_NEAR(0.095167236229178429, result.angle, 1e-11);
+        EXPECT_SOFT_EQ(0.55155207893668967, result.displacement);
+        EXPECT_SOFT_NEAR(0.095305171122713847, result.angle, 1e-11);
```
Are you sure these changes are needed? These are originally testing edge cases so I want to ensure we're still testing them.

mrguilima replied:
Checked, there is even good agreement between ORANGE and VecGeom surface and solids-model, both v1.x and v2.x.

sethrj commented on test/celeritas/optical/PrimaryGenerator.test.cc:
```diff
@@ -163,8 +167,10 @@ TEST_F(LArSpherePrimaryGeneratorTest, primary_generator)
+        constexpr int expected_steps = using_surface_vg ? 65536 : 68916;
+        constexpr int expected_step_iters = using_surface_vg ? 16 : 18;
+        EXPECT_EQ(expected_steps, result.steps);
+        EXPECT_EQ(expected_step_iters, result.step_iters);
```
Since these are just "regression" values and we're already ignoring the unusual case of single-precision real type, how about changing the `if` condition to ` ... && !CELERITAS_VECGEOM_SURFACE)` so we don't have to update this for multiple cases every time the optical loop changes.

sethrj commented on test/geocel/GenericGeoTestBase.t.hh:
```diff
@@ -390,12 +395,17 @@ auto GenericGeoTestBase<HP>::track(Real3 const& pos, Real3 const& dir)
+                if (using_surface_vg
+                    || (using_solids_vg && CELERITAS_VECGEOM_VERSION < 0x20000))
+                {
+                    // TODO: Understand why solids model v2.x fails here
```
What failure does it have?

sethrj commented on test/geocel/GeoTests.cc:
```diff
@@ -1979,6 +2034,10 @@ void ZnenvGeoTest::test_trace() const
         fixup_orange(*test_, ref, result, "World");
+        if (using_solids_vg)
+        {
+            GTEST_SKIP() << "FIME: Znenv VecGeom model construction failure.";
```
What failure? I think it works for the current vecgeom solids?

sethrj commented on src/geocel/vg/VecgeomTrackView.hh:
```diff
@@ -29,10 +29,8 @@
-#    include <VecGeom/navigation/BVHNavigator.h>
 #else
-#    include "detail/BVHNavigator.hh"
+#    include "detail/SolidsNavigator.hh"
```
This should mean that detail/BVHNavigator needs to be deleted, right?( It's basically been renamed)

sethrj reviewed (changes requested):
I know this has been a tedious struggle. Let's just get this over with...

mrguilima committed:
Merge branch 'develop' into vecgeom-2.0

sethrj committed:
REVERTME: disable other spack stuff

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vecgeom-2.0

mrguilima committed:
Address Seth's several suggestions

mrguilima committed:
Revert to using our BVHNavigator for v1.x solids model

mrguilima committed:
Put back relocation into cross_boundary()

mrguilima committed:
Fix some failing tests

mrguilima committed:
Fix some failing tests, skip some others for fixing later

sethrj committed:
Merge remote-tracking branch 'upstream/develop' into vecgeom-2.0

pre-commit-ci[bot] committed:
Apply fixes from pre-commit hooks: see detailed commit message →

This commit means that you must (please!) install pre-commit
on your development machine and run `pre-commit install --install-hooks`.
For more information, see
https://celeritas-project.github.io/celeritas/user/development/style.html#formatting

Autogenerated: https://pre-commit.ci

sethrj committed:
Revert timeout changes

sethrj committed:
Clean include warnings

sethrj committed:
Disable generator testing for vecgeom surface

sethrj committed:
fixup! Clean include warnings

sethrj committed:
Update expected failures for surface

sethrj committed:
Fix test when NDEBUG is off

sethrj committed:
Update tests

sethrj committed:
Restore CI jobs

sethrj committed:
Increase timeout for codecov

mrguilima committed:
Merge branch 'develop' into vecgeom-2.0

sethrj commented:
@pcanal do we need to add a "surface" vecgeom to codecov as well? 😅 now that it's clear we can build against both surface and solid in a single installation, I think we can include both in one build (but it'll take a bit of work).

sethrj commented:
@mrguilima I'm running before-and-after comparisons on GPU for select problems in the benchmark suite; if there are no substantial changes for the vg1 (and if the vg2 cases don't crash) I will merge. Thanks!

mrguilima commented:
> @mrguilima I'm running before-and-after comparisons on GPU for select problems in the benchmark suite; if there are no substantial changes for the vg1 (and if the vg2 cases don't crash) I will merge. Thanks!

I hope we're at a good point for merging.   
Thanks for the help to get to this point.

sethrj commented:
Results coming in now. Good news:
1. VecGeom 1.x performance is unaffected by the changes.
2. Some previously hanging vecgeom2 problems now finish in a reasonable number of step iterations.

Bad news:
1. Vecgeom solid  2.x is *three to five times slower* than 1.x.
2. Vecgeom surface 2.x still has a stuck track for CMS 2018. It's also 50% slower than 1.x

I think a major refactor of our use of vecgeom is in order. We need to avoid relocating when doing boundary checks, and we need to store the previous volume when crossing in order to correctly grab surfaces.  I'll do that in a future PR.

sethrj reviewed (approved):
The new navigators need work as discussed but the test results at least look reasonable. Thanks!
